wdtk-datamodel/src/main/java/org/wikidata/wdtk/datamodel/interfaces/TermedStatementDocument.java

```diff
@@ -0,0 +1,53 @@
+package org.wikidata.wdtk.datamodel.interfaces;
+
+/*
+ * #%L
+ * Wikidata Toolkit Data Model
+ * %%
+ * Copyright (C) 2014 - 2018 Wikidata Toolkit Developers
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This interface just joins {@link TermedDocument} and {@link StatementDocument}.
+ * 
+ * It is necessary to introduce this interface because the conflict between 
+ * the return types of the withRevisionId method in both interfaces needs to be resolved.
+ * @author antonin
+ *
+ */
+public interface TermedStatementDocument extends TermedDocument, StatementDocument {
```

Comment:
I would avoid to introduce this interface because it will introduce problems when we will want to implement support of lexemes that are going to be StatementDocument but not TermedDocument. A possible way, if we do not manage to get free of the withRevisionId method, would be to avoid overriding it in the TermedDocument and StatementDocument interfaces.

Reply:
In the current code we do need to override it in `TermedDocument` and `StatementDocument` (if we want `StatementUpdate` to return modified `StatementDocument` instances).

It's not clear to me why this interface is an issue - the lexemes are just not going to implement it, and that's fine (they will not be usable in the contexts where this interface is required). Classes like `StatementUpdate` should still be able to work on them.

Reply:
> the lexemes are just not going to implement it, and that's fine (they will not be usable in the contexts where this interface is required).
Yes, that's true.

I still don't like having such interface. But let's introduce it now and I'll try to figure out a way to drop it.
I also believe that we will want to get free of the hugly StatementUpdate and TermedStatementUpdate classes as soon as possible.

Reply:
yeah I want to refactor `StatementUpdate` and `TermedStatementUpdate` too, and introduce custom matching strategies for statements, qualifiers and references (the current matching mode is not always adequate)

Reply:
Great! So, let's keep this interface for now

Reply:
by the way what sort of structure are you thinking about for `StatementUpdate`? Beyond adding a builder for it?

Reply:
I was thinking about this workflow:

Item item = ...
item = item.withLabel("en", "Foo");
item = entityUpdater.saveItem(item, "my edit comment")

entityUpdater would do the diff with the item content in Wikidata for the relevant revisionId and pick the right edit action for you (and fail if there is an edit conflict).
But I am not sure that it is the best workflow. Have you an other in mind?

Reply:
hmm ok, so that `Item` class would keep track of both the initial `ItemDocument` and the new one. That would be very nice to use indeed! But AFAICT it's just about adding a new interface on top of the existing system (StatementUpdate would just not be manipulated by the user directly).

the refactoring I want to introduce is just to let users specify matching modes with each statement that they want to add. For instance, they might want to say "add this statement but only merge the reference with any existing statement if that existing statement has no reference yet", or "add this statement by matching only on mainsnak and merging qualifiers"… things like that.

Reply:
> hmm ok, so that Item class would keep track of both the initial ItemDocument and the new one. 

I would avoid that. It's too much responsibilities for this interface.

> the refactoring I want to introduce is just to let users specify matching modes with each statement that they want to add. For instance, they might want to say "add this statement but only merge the reference with any existing statement if that existing statement has no reference yet", or "add this statement by matching only on mainsnak and merging qualifiers"… things like that.

It looks great! We should probably move into this direction instead.

Reply:
> I would avoid that. It's too much responsibilities for this interface.

Yeah I did not mean putting both of these things in `ItemDocument` (I thought you had introduced a distinction between `Item` and `ItemDocument`).